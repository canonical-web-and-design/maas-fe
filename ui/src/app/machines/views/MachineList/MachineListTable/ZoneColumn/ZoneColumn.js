import { Spinner, Tooltip } from "@canonical/react-components";
import { useDispatch, useSelector } from "react-redux";
import { memo, useEffect, useState } from "react";
import PropTypes from "prop-types";

import baseURLs from "app/base/urls";
import { actions as machineActions } from "app/store/machine";
import machineSelectors from "app/store/machine/selectors";
import { NodeActions } from "app/store/types/node";
import zoneSelectors from "app/store/zone/selectors";
import { useToggleMenu } from "app/machines/hooks";
import DoubleRow from "app/base/components/DoubleRow";
import LegacyLink from "app/base/components/LegacyLink";

const getSpaces = (machine) => {
  if (machine.spaces.length > 1) {
    const sorted = [...machine.spaces].sort();
    return (
      <Tooltip position="btm-left" message={sorted.join("\n")}>
        <span data-test="spaces">{`${machine.spaces.length} spaces`}</span>
      </Tooltip>
    );
  }
  return (
    <span data-test="spaces" title={machine.spaces[0]}>
      {machine.spaces[0]}
    </span>
  );
};

export const ZoneColumn = ({ onToggleMenu, systemId }) => {
  const dispatch = useDispatch();
  const [updating, setUpdating] = useState(null);
  const machine = useSelector((state) =>
    machineSelectors.getById(state, systemId)
  );
  const zones = useSelector(zoneSelectors.all);
  const toggleMenu = useToggleMenu(onToggleMenu, systemId);
  let zoneLinks = zones.filter((zone) => zone.id !== machine.zone.id);
  if (machine.actions.includes(NodeActions.SET_ZONE)) {
    if (zoneLinks.length !== 0) {
      zoneLinks = zoneLinks.map((zone) => ({
        children: zone.name,
        onClick: () => {
          dispatch(machineActions.setZone({ systemId, zoneId: zone.id }));
          setUpdating(zone.id);
        },
      }));
    } else {
      zoneLinks = [{ children: "No other zones available", disabled: true }];
    }
  } else {
    zoneLinks = [
      { children: "Cannot change zone of this machine", disabled: true },
    ];
  }

  useEffect(() => {
    if (updating !== null && machine.zone.id === updating) {
      setUpdating(null);
    }
  }, [updating, machine.zone.id]);

  return (
    <DoubleRow
      menuLinks={onToggleMenu && zoneLinks}
      menuTitle="Change AZ:"
      onToggleMenu={toggleMenu}
      primary={
        <span data-test="zone">
          {updating !== null ? (
            <Spinner className="u-nudge-left--small" />
          ) : null}
          <LegacyLink
            className="p-link--soft"
            route={baseURLs.zone({ id: machine.zone.id })}
          >
            {machine.zone.name}
          </LegacyLink>
        </span>
      }
      primaryTitle={machine.zone.name}
      secondary={getSpaces(machine)}
    />
  );
};

ZoneColumn.propTypes = {
  onToggleMenu: PropTypes.func,
  systemId: PropTypes.string.isRequired,
};

export default memo(ZoneColumn);
