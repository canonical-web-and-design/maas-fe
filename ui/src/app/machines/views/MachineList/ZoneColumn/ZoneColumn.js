import { useSelector } from "react-redux";
import React from "react";
import PropTypes from "prop-types";

import { machine as machineSelectors } from "app/base/selectors";

const ZoneColumn = ({ systemId }) => {
  const machine = useSelector(state =>
    machineSelectors.getBySystemId(state, systemId)
  );

  const spaces =
    machine.spaces.length > 1
      ? `${machine.spaces.length} spaces`
      : machine.spaces[0];

  return (
    <div className="p-double-row">
      <div className="p-double-row__primary-row u-truncate">
        <span data-test="zone">{machine.zone.name}</span>
      </div>
      <div className="p-double-row__secondary-row u-truncate">
        <span data-test="spaces">{spaces}</span>
      </div>
    </div>
  );
};

ZoneColumn.propTypes = {
  systemId: PropTypes.string.isRequired
};

export default ZoneColumn;
