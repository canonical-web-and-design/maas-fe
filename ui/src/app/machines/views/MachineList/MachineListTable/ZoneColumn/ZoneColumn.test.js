import { act } from "react-dom/test-utils";
import { MemoryRouter } from "react-router-dom";
import { mount } from "enzyme";
import { Provider } from "react-redux";
import configureStore from "redux-mock-store";
import React from "react";

import { ZoneColumn } from "./ZoneColumn";

const mockStore = configureStore();

describe("ZoneColumn", () => {
  let state;
  beforeEach(() => {
    state = {
      config: {
        items: [],
      },
      machine: {
        errors: {},
        loading: false,
        loaded: true,
        items: [
          {
            system_id: "abc123",
            zone: { name: "zone-north", id: 0 },
            spaces: ["management"],
            actions: ["set-zone"],
          },
        ],
      },
      zone: {
        loaded: true,
        items: [
          {
            id: 0,
            name: "default",
          },
          {
            id: 1,
            name: "Backup",
          },
        ],
      },
    };
  });

  it("renders", () => {
    const store = mockStore(state);
    const wrapper = mount(
      <Provider store={store}>
        <MemoryRouter
          initialEntries={[{ pathname: "/machines", key: "testKey" }]}
        >
          <ZoneColumn onToggleMenu={jest.fn()} systemId="abc123" />
        </MemoryRouter>
      </Provider>
    );

    expect(wrapper.find("ZoneColumn")).toMatchSnapshot();
  });

  it("displays the zone name", () => {
    state.machine.items[0].zone.name = "zone-one";
    const store = mockStore(state);
    const wrapper = mount(
      <Provider store={store}>
        <MemoryRouter
          initialEntries={[{ pathname: "/machines", key: "testKey" }]}
        >
          <ZoneColumn onToggleMenu={jest.fn()} systemId="abc123" />
        </MemoryRouter>
      </Provider>
    );

    expect(wrapper.find('[data-test="zone"]').text()).toEqual("zone-one");
  });

  it("displays single space name", () => {
    state.machine.items[0].spaces = ["space1"];
    const store = mockStore(state);
    const wrapper = mount(
      <Provider store={store}>
        <MemoryRouter
          initialEntries={[{ pathname: "/machines", key: "testKey" }]}
        >
          <ZoneColumn onToggleMenu={jest.fn()} systemId="abc123" />
        </MemoryRouter>
      </Provider>
    );

    expect(wrapper.find('[data-test="spaces"]').text()).toEqual("space1");
  });

  it("displays spaces count for multiple spaces", () => {
    state.machine.items[0].spaces = ["space1", "space2"];
    const store = mockStore(state);
    const wrapper = mount(
      <Provider store={store}>
        <MemoryRouter
          initialEntries={[{ pathname: "/machines", key: "testKey" }]}
        >
          <ZoneColumn onToggleMenu={jest.fn()} systemId="abc123" />
        </MemoryRouter>
      </Provider>
    );

    expect(wrapper.find('[data-test="spaces"]').text()).toEqual("2 spaces");
  });

  it("displays a sorted Tooltip for multiple spaces", () => {
    state.machine.items[0].spaces = ["space2", "space1", "space3"];
    const store = mockStore(state);
    const wrapper = mount(
      <Provider store={store}>
        <MemoryRouter
          initialEntries={[{ pathname: "/machines", key: "testKey" }]}
        >
          <ZoneColumn onToggleMenu={jest.fn()} systemId="abc123" />
        </MemoryRouter>
      </Provider>
    );

    expect(wrapper.find("Tooltip").prop("message")).toEqual(
      "space1\nspace2\nspace3"
    );
  });

  it("displays a message if the machine cannot have its zone changed", () => {
    state.machine.items[0].actions = [];
    const store = mockStore(state);
    const wrapper = mount(
      <Provider store={store}>
        <MemoryRouter
          initialEntries={[{ pathname: "/machines", key: "testKey" }]}
        >
          <ZoneColumn onToggleMenu={jest.fn()} systemId="abc123" />
        </MemoryRouter>
      </Provider>
    );
    const items = wrapper.find("DoubleRow").prop("menuLinks");
    expect(items.length).toBe(1);
    expect(items[0]).toStrictEqual({
      children: "Cannot change zone of this machine",
      disabled: true,
    });
  });

  it("can change zones", () => {
    const store = mockStore(state);
    const wrapper = mount(
      <Provider store={store}>
        <MemoryRouter
          initialEntries={[{ pathname: "/machines", key: "testKey" }]}
        >
          <ZoneColumn onToggleMenu={jest.fn()} systemId="abc123" />
        </MemoryRouter>
      </Provider>
    );
    act(() => {
      wrapper.find("DoubleRow").prop("menuLinks")[0].onClick();
    });
    expect(
      store.getActions().find((action) => action.type === "SET_MACHINE_ZONE")
    ).toEqual({
      type: "SET_MACHINE_ZONE",
      meta: {
        model: "machine",
        method: "action",
      },
      payload: {
        params: {
          action: "set-zone",
          extra: {
            zone_id: 1,
          },
          system_id: "abc123",
        },
      },
    });
  });

  it("shows a spinner when changing zones", () => {
    const store = mockStore(state);
    const wrapper = mount(
      <Provider store={store}>
        <MemoryRouter
          initialEntries={[{ pathname: "/machines", key: "testKey" }]}
        >
          <ZoneColumn onToggleMenu={jest.fn()} systemId="abc123" />
        </MemoryRouter>
      </Provider>
    );
    expect(wrapper.find("Spinner").exists()).toBe(false);
    act(() => {
      wrapper.find("DoubleRow").prop("menuLinks")[0].onClick();
    });
    wrapper.update();
    expect(wrapper.find("Spinner").exists()).toBe(true);
  });

  it("does not render table menu if onToggleMenu not provided", () => {
    const store = mockStore(state);
    const wrapper = mount(
      <Provider store={store}>
        <MemoryRouter
          initialEntries={[{ pathname: "/machines", key: "testKey" }]}
        >
          <ZoneColumn systemId="abc123" />
        </MemoryRouter>
      </Provider>
    );

    expect(wrapper.find("TableMenu").exists()).toBe(false);
  });
});
