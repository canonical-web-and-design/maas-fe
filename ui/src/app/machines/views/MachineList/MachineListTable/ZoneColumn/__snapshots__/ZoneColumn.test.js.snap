// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ZoneColumn renders 1`] = `
<ZoneColumn
  onToggleMenu={[MockFunction]}
  systemId="abc123"
>
  <DoubleRow
    menuLinks={
      Array [
        Object {
          "children": "Backup",
          "onClick": [Function],
        },
      ]
    }
    menuTitle="Change AZ:"
    onToggleMenu={[Function]}
    primary={
      <span
        data-test="zone"
      >
        <a
          className="p-link--soft"
          href="/MAAS/l/zone/0"
        >
          zone-north
        </a>
      </span>
    }
    primaryTitle="zone-north"
    secondary={
      <span
        data-test="spaces"
        title="management"
      >
        management
      </span>
    }
  >
    <div
      className="p-double-row"
    >
      <div
        className="p-double-row__rows-container"
      >
        <div
          className="p-double-row__primary-row"
        >
          <div
            className="p-double-row__primary-row-text u-truncate"
            title="zone-north"
          >
            <span
              data-test="zone"
            >
              <a
                className="p-link--soft"
                href="/MAAS/l/zone/0"
              >
                zone-north
              </a>
            </span>
          </div>
          <TableMenu
            links={
              Array [
                Object {
                  "children": "Backup",
                  "onClick": [Function],
                },
              ]
            }
            onToggleMenu={[Function]}
            positionNode={
              Object {
                "current": <div
                  class="p-double-row__primary-row"
                >
                  <div
                    class="p-double-row__primary-row-text u-truncate"
                    title="zone-north"
                  >
                    <span
                      data-test="zone"
                    >
                      <a
                        class="p-link--soft"
                        href="/MAAS/l/zone/0"
                      >
                        zone-north
                      </a>
                    </span>
                  </div>
                  <span
                    class="p-table-menu p-contextual-menu--left"
                  >
                    <button
                      aria-controls="Uakgb_J5m9g-0JDMbcJqLJ"
                      aria-expanded="false"
                      aria-haspopup="true"
                      class="p-button--base has-icon p-contextual-menu__toggle u-no-margin--bottom p-table-menu__toggle"
                    >
                      <i
                        class="p-icon--contextual-menu p-contextual-menu__indicator"
                      />
                    </button>
                  </span>
                </div>,
              }
            }
            title="Change AZ:"
          >
            <ContextualMenu
              className="p-table-menu"
              hasToggleIcon={true}
              links={
                Array [
                  "Change AZ:",
                  Object {
                    "children": "Backup",
                    "onClick": [Function],
                  },
                ]
              }
              onToggleMenu={[Function]}
              position="left"
              positionNode={
                Object {
                  "current": <div
                    class="p-double-row__primary-row"
                  >
                    <div
                      class="p-double-row__primary-row-text u-truncate"
                      title="zone-north"
                    >
                      <span
                        data-test="zone"
                      >
                        <a
                          class="p-link--soft"
                          href="/MAAS/l/zone/0"
                        >
                          zone-north
                        </a>
                      </span>
                    </div>
                    <span
                      class="p-table-menu p-contextual-menu--left"
                    >
                      <button
                        aria-controls="Uakgb_J5m9g-0JDMbcJqLJ"
                        aria-expanded="false"
                        aria-haspopup="true"
                        class="p-button--base has-icon p-contextual-menu__toggle u-no-margin--bottom p-table-menu__toggle"
                      >
                        <i
                          class="p-icon--contextual-menu p-contextual-menu__indicator"
                        />
                      </button>
                    </span>
                  </div>,
                }
              }
              toggleAppearance="base"
              toggleClassName="u-no-margin--bottom p-table-menu__toggle"
            >
              <span
                className="p-table-menu p-contextual-menu--left"
              >
                <Button
                  appearance="base"
                  aria-controls="Uakgb_J5m9g-0JDMbcJqLJ"
                  aria-expanded="false"
                  aria-haspopup="true"
                  className="p-contextual-menu__toggle u-no-margin--bottom p-table-menu__toggle"
                  hasIcon={true}
                  onClick={[Function]}
                >
                  <button
                    aria-controls="Uakgb_J5m9g-0JDMbcJqLJ"
                    aria-expanded="false"
                    aria-haspopup="true"
                    className="p-button--base has-icon p-contextual-menu__toggle u-no-margin--bottom p-table-menu__toggle"
                    onClick={[Function]}
                  >
                    <i
                      className="p-icon--contextual-menu p-contextual-menu__indicator"
                    />
                  </button>
                </Button>
              </span>
            </ContextualMenu>
          </TableMenu>
        </div>
        <div
          className="p-double-row__secondary-row u-truncate"
        >
          <span
            data-test="spaces"
            title="management"
          >
            management
          </span>
        </div>
      </div>
    </div>
  </DoubleRow>
</ZoneColumn>
`;
